crates/oci-distribution/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+#[macro_use]
+extern crate serde;
+
+use hyperx::header::Header;
+use oauth2::{Client as OauthClient, StandardToken};
+use serde::Deserialize;
+use std::convert::TryFrom;
+use url::Url;
+use www_authenticate::{Challenge, ChallengeFields, RawChallenge, WwwAuthenticate};
+
+use crate::errors::*;
+use crate::reference::Reference;
+
+const OCI_VERSION_KEY: &str = "Docker-Distribution-Api-Version";
+
+pub mod errors;
+pub mod reference;
+
+type OciResult<T> = Result<T, anyhow::Error>;
+
+struct Client {
+    token: Option<StandardToken>,
+}
+
+impl Default for Client {
+    fn default() -> Self {
+        Client { token: None }
+    }
+}
+
+impl Client {
+    /// According to the v2 specification, 200 and 401 error codes MUST return the
+    /// version. It appears that any other response code should be deemed non-v2.
+    ///
+    /// For this implementation, it will return v2 or an error result. If the error is a
+    /// `reqwest` error, the request itself failed. All other error messages mean that
+    /// v2 is not supported.
+    pub async fn version(&self, host: String) -> OciResult<String> {
+        let url = format!("https://{}/v2/", host);
+        let res = reqwest::get(&url).await?;
+        let disthdr = res.headers().get(OCI_VERSION_KEY);
+        let version = disthdr
+            .ok_or_else(|| anyhow::format_err!("no header v2 found"))?
+            .to_str()?
+            .to_owned();
+        Ok(version)
+    }
+
+    /// Perform an OAuth v2 auth request if necessary.
+    ///
+    /// This performs authorization and then stores the token internally to be used
+    /// on other requests.
+    pub async fn auth(&mut self, host: String, secret: Option<String>) -> OciResult<()> {
+        let cli = reqwest::Client::new();
+        // The version request will tell us where to go.
+        let url = format!("https://{}/v2/", host);
+        let res = cli.get(&url).send().await?;
+        let disthdr = res.headers().get(reqwest::header::WWW_AUTHENTICATE);
+        if disthdr.is_none() {
```

Comment:
Maybe a `match` or an `if let Some(..)` is more appropriate here to avoid the unwrap a few lines below?

Reply:
This is probably my Go-brain speaking, but to avoid apparent cyclomatic complexity, there are cases where we want to return early without increasing the indent. What is the right way to do that in Rust? Both `match` and `if let Some()` would increase the indent (and thus the apparent cyclomatic complexity). `if let None` seems about the same as `if is_none()`.